display_list/effects/dl_color_source.cc

```diff
@@ -157,23 +181,71 @@ bool DlGradientColorSourceBase::base_equals_(
                  stop_count_ * sizeof(stops()[0])) == 0);
 }
 
-void DlGradientColorSourceBase::store_color_stops(void* pod,
-                                                  const DlColor* color_data,
-                                                  const float* stop_data) {
+namespace {
+template <typename DlColorIt>
+void do_store_color_stops(void* pod,
+                          DlColorIt color_data_argb_begin,
+                          DlColorIt color_data_argb_end,
+                          const float* stop_data) {
   DlColor* color_storage = reinterpret_cast<DlColor*>(pod);
-  memcpy(color_storage, color_data, stop_count_ * sizeof(*color_data));
-  float* stop_storage = reinterpret_cast<float*>(color_storage + stop_count_);
+  uint32_t stop_count = 0;
+  while (color_data_argb_begin < color_data_argb_end) {
+    *color_storage++ = *color_data_argb_begin++;
+    stop_count += 1;
+  }
+  float* stop_storage = reinterpret_cast<float*>(color_storage);
   if (stop_data) {
-    memcpy(stop_storage, stop_data, stop_count_ * sizeof(*stop_data));
+    memcpy(stop_storage, stop_data, stop_count * sizeof(*stop_data));
   } else {
-    float div = stop_count_ - 1;
+    float div = stop_count - 1;
     if (div <= 0) {
       div = 1;
     }
-    for (uint32_t i = 0; i < stop_count_; i++) {
+    for (uint32_t i = 0; i < stop_count; i++) {
       stop_storage[i] = i / div;
     }
   }
 }
 
+class DlScalerToDlColorIt {
```

Comment:
Til how iterators work in C++.

Maybe `DlColorConversionIterator`?

Reply:
> Til how iterators work in C++.

Yea, just think of using operator overloading to make a class work like a pointer and use templates to accept anything that works like a pointer.

> Maybe DlColorConversionIterator?

done